### PR TITLE
Fixed  issue 1745

### PR DESCRIPTION
**Pull request by ZhangHanDong**
This PR fixes issue [#1745](https://github.com/ratatui/ratatui/issues/1745) where Chinese characters and other multi-width characters were being rendered with additional unwanted whitespace between them. The problem occurred in the set_stringn method when handling characters that occupy multiple cells in the terminal.

The fix ensures proper handling of multi-width characters by explicitly resetting cells covered by these characters and advancing the cursor position by the exact width of each character. This prevents the insertion of unexpected whitespace between characters and maintains correct alignment.

A comprehensive test suite has been added to verify both pure Chinese text and mixed English/Chinese text rendering, confirming that multi-width characters occupy the correct number of cells without additional spacing.

**However, since I was unable to reproduce the error in issue 1745 locally, I tried adding Chinese to the todo list, and it seems to work fine.**

Screenshot 2025-04-05 at 17 05 29

### CODE REVIEW

File: ratatui-core/src/buffer/buffer.rs [hunk 1]
```diff
@@ -345,19 +345,25 @@ impl Buffer {
...
             self[(x, y)].set_symbol(symbol).set_style(style);
-            let next_symbol = x + width;
-            x += 1;
-            // Reset following cells if multi-width (they would be hidden by the grapheme),
-            while x < next_symbol {
-                self[(x, y)].reset();
-                x += 1;
+            // For each grapheme, advance x by exactly its width
+            // This ensures that multi-width characters like Chinese characters
+            // are properly handled without adding extra whitespace
```
Comment (joshka) [inline ratatui-core/src/buffer/buffer.rs:360]: nit: wrap comments at 100 chars

File: ratatui-core/src/buffer/buffer.rs [hunk 2]
```diff
@@ -345,19 +345,25 @@ impl Buffer {
...
+            for i in 1..width {
+                if x + i < self.area.right() {
+                    self[(x + i, y)].reset();
+                }
             }
+            x += width;
```
Comment (joshka) [inline ratatui-core/src/buffer/buffer.rs:366]: This is definitely better, but it's not clear that the if check is required (the purpose of the map_while above is to exclude characters which would make `x + width >= area.right()`

File: ratatui-core/src/buffer/buffer.rs [hunk 3]
```diff
@@ -655,6 +661,47 @@ mod tests {
         assert_eq!(result, expected);
     }
 
+    #[test]
+    fn chinese_characters_rendering() {
```
Comment (joshka) [inline ratatui-core/src/buffer/buffer.rs:665]: I checked this test against the current code and it succeeds already.

**Review by joshka — COMMENTED** [review]
It's not clear what's actually changed here that would fix issue 1745. From my reading, of this it seems like there's actually no logical change here.

I suspect that the problems would be around the edge cases that occur when a wide character is at the end of line.

The way to show that the problem is fixed is to create a small unit test that actually triggers the problem, and to see that unit test go from failing to succeeding with the change. It's not clear that's happening here.

**Comment by orhun** [comment]
I agree that adding chinese characters to the examples is a good idea (and I like the test suite) but I'm not so sure if this PR touches any existing rendering issues...

File: ratatui-core/src/buffer/buffer.rs (hunk 3 above)
Comment (ZhangHanDong) [reply ratatui-core/src/buffer/buffer.rs:665]: I’ve added more tests locally and found that the current main branch code will fail when executing these tests. I’ll submit the modifications after conducting more tests and pinpointing some issues.

**Comment by ZhangHanDong** [comment]
@joshka 

I added two new tests that fail in the buffer.rs file of the main branch. 

```rust
    #[test]
    fn chinese_at_buffer_boundary() {
        let area = Rect::new(0, 0, 5, 1);
        let mut buffer = Buffer::empty(area);

        // Should fit exactly: "a" + "你" + "b"
        buffer.set_string(0, 0, "a你b", Style::default());

        let contents = buffer.content.iter().map(Cell::symbol).collect::<Vec<_>>();
        assert_eq!(contents[0], "a");
        assert_eq!(contents[1], "你");
        assert_eq!(contents[2], " "); // Should be reset/empty
        assert_eq!(contents[3], "b");

        // Now try rendering where the last character would be cut off
        buffer = Buffer::empty(area);
        buffer.set_string(0, 0, "a你好", Style::default());

        let contents = buffer.content.iter().map(Cell::symbol).collect::<Vec<_>>();
        assert_eq!(contents[0], "a");
        assert_eq!(contents[1], "你");
        assert_eq!(contents[2], " "); // Should be reset/empty
        // Character "好" should not appear as it doesn't fit
        assert_eq!(contents[3], " ");
        assert_eq!(contents[4], " ");
    }

    #[test]
    fn visual_chinese_formatting() {
        // Create a buffer with border characters and Chinese text
        let area = Rect::new(0, 0, 14, 3);
        let mut buffer = Buffer::empty(area);

        // Draw a box
        buffer.set_string(0, 0, "┌────────┐", Style::default());
        buffer.set_string(0, 1, "│        │", Style::default());
        buffer.set_string(0, 2, "└────────┘", Style::default());

        // Put Chinese characters inside the box
        buffer.set_string(2, 1, "你好世界", Style::default());

        // The entire buffer content should look correct with proper alignment
        let expected = [
            "┌────────┐",
            "│你好世界│", // 4 Chinese characters taking 8 spaces
            "└────────┘"
        ];

        // Verify each row
        for (i, row) in expected.iter().enumerate() {
            let mut line = String::new();
            for j in 0..area.width {
                line.push_str(buffer[(j, i as u16)].symbol());
            }
            assert_eq!(line, *row);
        }
    }
```

The test results are as follows:

```
---- buffer::buffer::tests::chinese_at_buffer_boundary stdout ----

thread 'buffer::buffer::tests::chinese_at_buffer_boundary' panicked at ratatui-core/src/buffer/buffer.rs:1396:9:
assertion `left == right` failed
  left: "好"
 right: " "
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- buffer::buffer::tests::visual_chinese_formatting stdout ----

thread 'buffer::buffer::tests::visual_chinese_formatting' panicked at ratatui-core/src/buffer/buffer.rs:1449:13:
assertion `left == right` failed
  left: "┌────────┐    "
 right: "┌────────┐"


failures:
    buffer::buffer::tests::chinese_at_buffer_boundary
    buffer::buffer::tests::visual_chinese_formatting
```

**Comment by ZhangHanDong** [comment]
I'm adding more multi-byte related test cases to locate and identify issues. 

Screenshot 2025-04-07 at 11 12 18

I’ll send a new update if there’s a final fix.

 @orhun @joshka

**Comment by joshka** [comment]
marking as a draft for now until it's resolved / questions are answered.
